src/UCTNode.cpp

```diff
@@ -92,8 +93,20 @@ bool UCTNode::create_children(Network & network,
             legal_sum += raw_netlist.policy[i];
         }
     }
-    nodelist.emplace_back(raw_netlist.policy_pass, FastBoard::PASS);
-    legal_sum += raw_netlist.policy_pass;
+
+    // Don't consider pass moves if they would be immediately losing and we "shouldn't"
+    // accept that; see issue #2273. --dumbpass disables this heuristic.
+    float final_score_for_player = state.final_score();
+    if (to_move == FastBoard::WHITE) final_score_for_player = -final_score_for_player;
+    bool suppress_passes =
+        winrate_for_player >= 0.75     // we should be winning
```

Comment:
Is there any reason to check the winrate now that number of pass condition was removed? I don't see the point in allowing pass at low winrates if it's losing.

Reply:
It's common for a human game to end with two consecutive passes even though one of the players knows that they're going to lose. I confess I haven't checked how often that actually happens in LZ games rather than LZ resigning once it's sure it's lost. I think I've seen no-resign selfplay games, at least, end with two consecutive passes.

I'm happy to defer to others who actually know what they're doing here. Would anything bad happen if LZ literally never passed when ending the game would result in a loss?

Reply:
Passing loses instantly, but any other move would have possibly non-zero chance to win. In self-play I don't think the losing player ever passes second, but it can pass first due to tree search averaging giving pass greater than zero winning chance. Self-play uses dumbpass option so this wouldn't affect it.

Removing this condition should be beneficial for handicap play since no playouts are spent for unnecessary passing moves. In match play resign should be enabled and the engine would resign before passing is relevant.

Reply:
The winrate check does serve one other possibly useful purpose: it avoids bothering with the final-score check in some cases (i.e., those where the player to move is not comfortably ahead). So if the cost of the final-score check isn't entirely negligible -- as in another comment you suggested it might be -- then the winrate check brings some benefit by skipping roughly half the final-score checks (if one player is comfortably ahead) or most of them (if the position is roughly equal).

On the other hand, if there are many nodes where winrate isn't high, the policy on passing is highish, but passing would lose, then the search will be a bit more efficient for the reason you mention.

It's not obvious to me how these considerations balance out. I suspect there's very little difference overall.